libopeniscsiusr/iface.c

```diff
@@ -268,6 +267,52 @@ int _iscsi_iface_get_from_sysfs(struct iscsi_context *ctx, uint32_t host_id,
 	return rc;
 }
 
+/* create all ifaces for a host from sysfs */
+int _iscsi_ifaces_get_from_sysfs(struct iscsi_context *ctx, uint32_t host_id,
+				 struct iscsi_iface ***ifaces, uint32_t *iface_count)
+{
+	int rc = LIBISCSI_OK;
+	char **iface_kern_ids = NULL;
+
+	assert(ctx != NULL);
+	assert(ifaces != NULL);
+
+	*ifaces = NULL;
+	*iface_count = 0;
+
+	_good(_iscsi_iface_kern_ids_of_host_id(ctx, host_id, &iface_kern_ids, iface_count),
+	      rc, out);
+	if (*iface_count > 0) {
+		*ifaces = (struct iscsi_iface **) calloc(*iface_count,
+							 sizeof(struct iscsi_iface *));
+		_alloc_null_check(ctx, *ifaces, rc, out);
+		for (uint32_t i = 0; i < *iface_count; i++) {
+			_good(_iscsi_iface_get_from_sysfs(ctx, host_id, 0,
+					iface_kern_ids[i], &(*ifaces)[i]), rc, out);
+		}
+	} else {
+		/* if there's no iface exported in sysfs,
+		 * we should still be able to create one record per host */
+		*ifaces = (struct iscsi_iface **) calloc(1, sizeof(struct iscsi_iface *));
+		_alloc_null_check(ctx, *ifaces, rc, out);
+		_good(_iscsi_iface_get_from_sysfs(ctx, host_id, 0, NULL, &(*ifaces)[0]), rc, out);
+		*iface_count = 1;
```

Comment:
Memory leak here.
If `_iscsi_iface_get_from_sysfs()` failed, `iface_count` is `0`, the memory hold by `*ifaces` will not be freed.

Reply:
good find, I'll move setting *iface_count = 1 up a line, so it's between the _alloc_null_check and the call to _iscsi_iface_get_from_sysfs